File: AlgorithmLcr/Try2-Known component types, quantities and parameters/netgraph_id/graph.py
"""Multigraph model of a 2-terminal RLC network (DESIGN.md section 3).

A network is a connected undirected multigraph:
  * nodes 0 and 1 are the two port terminals, nodes 2..V-1 are internal
    junctions;
  * edges are components -- parallel edges (same node pair) are legal and
    are exactly "components in parallel"; self loops never occur (a
    component whose leads are joined to the same node carries no external
    current);
  * every edge is an independent object (never keyed by its node pair).

Two representations (DESIGN.md section 3.2):
  * structure layer: symmetric multiplicity matrix M[i][j] = m_ij (zero
    diagonal, sum = E) stored as a flat tuple over the canonical slot list;
    this carries enumeration / canonicalization / dedup;
  * assignment layer: an edge list assigning concrete components to the
    edge instances of a structure.

Reduction rule R0 (dead-part exclusion, DESIGN.md section 4.3): any
connected component of G - c (c any node) that contains no terminal is
electrically dead -- summing KCL over its nodes shows the net current
flowing into it through c is identically zero -- so structures containing
such parts are excluded from the enumeration.
"""

from __future__ import annotations

from dataclasses import dataclass
from itertools import permutations
from typing import Iterator

# ---------------------------------------------------------------------------
# slots and multiplicity vectors
# ---------------------------------------------------------------------------

_SLOT_CACHE: dict[int, list[tuple[int, int]]] = {}
_SLOT_INDEX_CACHE: dict[int, dict[tuple[int, int], int]] = {}


def slot_list(V: int) -> list[tuple[int, int]]:
    """Canonical slot list [(0,1),(0,2),...,(1,2),...,(V-2,V-1)]."""
    if V not in _SLOT_CACHE:
        _SLOT_CACHE[V] = [(i, j) for i in range(V) for j in range(i + 1, V)]
    return _SLOT_CACHE[V]


def slot_index(V: int) -> dict[tuple[int, int], int]:
    if V not in _SLOT_INDEX_CACHE:
        _SLOT_INDEX_CACHE[V] = {ij: k for k, ij in enumerate(slot_list(V))}
    return _SLOT_INDEX_CACHE[V]


def n_slots(V: int) -> int:
    return V * (V - 1) // 2


def empty_mult(V: int) -> list[int]:
    return [0] * n_slots(V)


def mult_degree(V: int, mult) -> list[int]:
    """Multigraph degree (parallel edges counted) of every node."""
    deg = [0] * V
    for (i, j), m in zip(slot_list(V), mult):
        deg[i] += m
        deg[j] += m
    return deg


# ---------------------------------------------------------------------------
# connectivity and the R0 dead-part rule
# ---------------------------------------------------------------------------

def is_connected(V: int, mult) -> bool:
    """True iff all V nodes lie in one connected component (union-find)."""
    parent = list(range(V))

    def find(a: int) -> int:
        while parent[a] != a:
            parent[a] = parent[parent[a]]
            a = parent[a]
        return a

    for (i, j), m in zip(slot_list(V), mult):
        if m > 0:
            ri, rj = find(i), find(j)
            if ri != rj:
                parent[ri] = rj
    root = find(0)
    return all(find(v) == root for v in range(V))


def has_dead_part(V: int, mult) -> bool:
    """R0 test: does some connected component of G - c contain no terminal?

    Such a component hangs off the rest of the network through the single
    node c; summing KCL over its nodes proves the net current entering it
    is identically zero, so it is electrically invisible (DESIGN.md 4.3).
    A single pass suffices: any dead part P attached only via c must be a
    *full* component of G - c, because a proper sub-part of a component of
    G - c would have an additional attachment inside that component.
    """
    if V <= 2:
        return False
    adj: list[list[int]] = [[] for _ in range(V)]
    for (i, j), m in zip(slot_list(V), mult):
        if m > 0:
            adj[i].append(j)
            adj[j].append(i)
    for c in range(V):
        seen = [False] * V
        for start in range(V):
            if start == c or seen[start]:
                continue
            # BFS the component of start in G - c
            stack = [start]
            seen[start] = True
            has_terminal = False
            while stack:
                x = stack.pop()
                if x in (0, 1):
                    has_terminal = True
                for y in adj[x]:
                    if y != c and not seen[y]:
                        seen[y] = True
                        stack.append(y)
            if not has_terminal:
                return True
    return False


def structure_ok(V: int, mult, *, allow_dead: bool = False) -> bool:
    """Admissibility of a uniform structure (connectivity + R0)."""
    if not is_connected(V, mult):
        return False
    if not allow_dead and has_dead_part(V, mult):
        return False
    return True


# ---------------------------------------------------------------------------
# node permutations, canonical form, automorphisms
# ---------------------------------------------------------------------------

def perm_group(V: int) -> Iterator[tuple[int, ...]]:
    """The relabeling group G = {terminal swap} x Sym(internal nodes).

    Node 0/1 are distinguished as the port pair but exchanging them leaves
    the driving-point impedance unchanged, so both labellings are identified;
    internal nodes 2..V-1 are freely permutable.
    """
    internal = range(2, V)
    for swap in (False, True):
        for pi in permutations(internal):
            p = [0] * V
            p[0], p[1] = (1, 0) if swap else (0, 1)
            for k, v in enumerate(pi):
                p[v] = k + 2
            yield tuple(p)


def permute_mult(V: int, mult, p) -> tuple[int, ...]:
    """Relabel nodes by p (image of node i is p[i]) and rebuild the slot vector."""
    si = slot_index(V)
    out = [0] * n_slots(V)
    for (i, j), m in zip(slot_list(V), mult):
        a, b = p[i], p[j]
        if a > b:
            a, b = b, a
        out[si[(a, b)]] += m
    return tuple(out)


def canonical_mult(V: int, mult) -> tuple[int, ...]:
    """Canonical representative = lexicographic minimum over G."""
    return min(permute_mult(V, mult, p) for p in perm_group(V))


def structure_automorphisms(V: int, mult) -> tuple[tuple[int, ...], ...]:
    """All elements of G that fix the (canonical) mult vector."""
    return tuple(p for p in perm_group(V) if permute_mult(V, mult, p) == tuple(mult))


# ---------------------------------------------------------------------------
# structure container
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class Structure:
    """One uniform-edge 2-terminal multigraph structure (canonical form).

    mult[k] = number of parallel edges in slot_list(V)[k]; aut = node
    permutations (from G) fixing the structure, used to dedup component
    assignments.
    """

    V: int
    mult: tuple[int, ...]
    aut: tuple[tuple[int, ...], ...]

    @property
    def n_edges(self) -> int:
        return sum(self.mult)

    @property
    def n_internal(self) -> int:
        return self.V - 2

    @property
    def key(self) -> tuple:
        return (self.V, self.mult)

    def degrees(self) -> list[int]:
        return mult_degree(self.V, self.mult)

    def slot_of_instances(self) -> tuple[int, ...]:
        """Edge-instance order: slots in canonical order, instances within a
        slot consecutive.  Length = E."""
        out: list[int] = []
        for k, m in enumerate(self.mult):
            out.extend([k] * m)
        return tuple(out)

    def instance_pairs(self) -> list[tuple[int, int]]:
        slots = slot_list(self.V)
        soi = self.slot_of_instances()
        return [slots[k] for k in soi]

    def serialize(self) -> str:
        return f"V{self.V}:" + ",".join(str(m) for m in self.mult)


def make_structure(V: int, mult, *, canonicalize: bool = True) -> Structure:
    """Build a Structure, optionally mapping to its canonical form first."""
    mult = tuple(mult)
    if len(mult) != n_slots(V):
        raise ValueError(f"mult has {len(mult)} slots, expected {n_slots(V)}")
    if canonicalize:
        mult = canonical_mult(V, mult)
    return Structure(V=V, mult=mult, aut=structure_automorphisms(V, mult))


# ---------------------------------------------------------------------------
# networks (structure + component assignment) and SP recognition
# ---------------------------------------------------------------------------

def permute_slot_keys(V: int, keys_per_slot: tuple, p) -> tuple:
    """Move per-slot content (sorted component-key tuples) along a node
    permutation; used to canonicalize component assignments."""
    si = slot_index(V)
    out: list[list] = [[] for _ in range(len(keys_per_slot))]
    for (i, j), keys in zip(slot_list(V), keys_per_slot):
        if not keys:
            continue
        a, b = p[i], p[j]
        if a > b:
            a, b = b, a
        out[si[(a, b)]].extend(keys)
    return tuple(tuple(sorted(g)) for g in out)


@dataclass(frozen=True)
class Network:
    """A candidate network: structure + assignment of components to edges.

    assign[t] = index into the canonical (sorted) component tuple, for edge
    instance t (instance order = Structure.slot_of_instances()).
    """

    structure: Structure
    assign: tuple[int, ...]

    def serialize(self, comp_keys: list[tuple]) -> tuple:
        """Canonical serialization: per-slot sorted component keys, minimized
        over the structure automorphism group.  Two networks of the same
        structure are electrically-identical-as-wirings iff equal serial."""
        V = self.structure.V
        soi = self.structure.slot_of_instances()
        per_slot: list[list[tuple]] = [[] for _ in range(n_slots(V))]
        for t, comp_idx in enumerate(self.assign):
            per_slot[soi[t]].append(comp_keys[comp_idx])
        base = tuple(tuple(sorted(g)) for g in per_slot)
        if len(self.structure.aut) <= 1:
            return base
        return min(permute_slot_keys(V, base, p) for p in self.structure.aut)


def is_series_parallel(V: int, mult) -> bool:
    """Valdes-Tarjan-Puech two-terminal series-parallel recognition.

    Repeatedly (a) merge parallel edges, (b) contract non-terminal degree-2
    nodes (series); the graph is two-terminal SP iff it reduces to the
    single edge (0,1).
    """
    if V < 2:
        return False
    counts = {(i, j): m for (i, j), m in zip(slot_list(V), mult) if m > 0}

    alive = set(range(V))
    while True:
        # (a) parallel reduction: collapse every multi-edge to a single edge
        for key in [k for k, m in counts.items() if m > 1]:
            counts[key] = 1
        # (b) one series reduction per pass (indices stay consistent);
        #     reductions are confluent, so the order does not matter
        deg = [0] * V
        inc: list[list[tuple[int, int]]] = [[] for _ in range(V)]
        for (a, b), m in counts.items():
            deg[a] += m
            deg[b] += m
            for _ in range(m):
                inc[a].append((a, b))
                inc[b].append((a, b))
        target = next((x for x in alive
                       if x not in (0, 1) and deg[x] == 2), None)
        if target is None:
            break
        x = target
        (a1, b1), (a2, b2) = inc[x][0], inc[x][1]
        for key in ((a1, b1), (a2, b2)):
            counts[key] -= 1
        for key in ((a1, b1), (a2, b2)):
            if counts.get(key, 0) <= 0:
                counts.pop(key, None)
        alive.discard(x)
        others = {a1, b1, a2, b2} - {x}
        if len(others) == 2:
            u, v = sorted(others)
            key = (u, v)
            counts[key] = counts.get(key, 0) + 1
        # len(others) == 1: both edges went to the same partner -> dead pair
    return counts.get((0, 1), 0) == 1 and len(counts) == 1

File: AlgorithmLcr/Try2-Known component types, quantities and parameters/cppversion/src/graph.cpp
#include "graph.hpp"

#include <algorithm>
#include <numeric>
#include <sstream>
#include <stdexcept>

namespace ng {

namespace {

// Union-find with path halving over 0..V-1.
struct Dsu {
    std::vector<int> parent;
    explicit Dsu(int n) : parent((size_t)n) { std::iota(parent.begin(), parent.end(), 0); }
    int find(int a) {
        while (parent[a] != a) {
            parent[a] = parent[parent[a]];
            a = parent[a];
        }
        return a;
    }
    void unite(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra != rb) parent[ra] = rb;
    }
};

std::vector<std::pair<int, int>>& slotListMutable(int V) {
    static thread_local std::map<int, std::vector<std::pair<int, int>>> cache;
    auto it = cache.find(V);
    if (it == cache.end()) {
        std::vector<std::pair<int, int>> slots;
        for (int i = 0; i < V; ++i)
            for (int j = i + 1; j < V; ++j) slots.push_back({i, j});
        it = cache.emplace(V, std::move(slots)).first;
    }
    return it->second;
}

}  // namespace

const std::vector<std::pair<int, int>>& slotList(int V) { return slotListMutable(V); }

int nSlots(int V) { return V * (V - 1) / 2; }

int slotIndex(int V, int i, int j) {
    // row-major upper triangle: rows 0..i-1 hold V-1-a slots each, then
    // offset (j - i - 1) inside row i.
    return i * (V - 1) - i * (i - 1) / 2 + (j - i - 1);
}

std::vector<int> emptyMult(int V) { return std::vector<int>((size_t)nSlots(V), 0); }

std::vector<int> multDegree(int V, const std::vector<int>& mult) {
    std::vector<int> deg((size_t)V, 0);
    const auto& slots = slotList(V);
    for (size_t k = 0; k < slots.size(); ++k) {
        deg[slots[k].first] += mult[k];
        deg[slots[k].second] += mult[k];
    }
    return deg;
}

bool isConnected(int V, const std::vector<int>& mult) {
    Dsu dsu(V);
    const auto& slots = slotList(V);
    for (size_t k = 0; k < slots.size(); ++k) {
        if (mult[k] > 0) dsu.unite(slots[k].first, slots[k].second);
    }
    int root = dsu.find(0);
    for (int v = 0; v < V; ++v)
        if (dsu.find(v) != root) return false;
    return true;
}

bool hasDeadPart(int V, const std::vector<int>& mult) {
    if (V <= 2) return false;
    std::vector<std::vector<int>> adj((size_t)V);
    const auto& slots = slotList(V);
    for (size_t k = 0; k < slots.size(); ++k) {
        if (mult[k] > 0) {
            adj[slots[k].first].push_back(slots[k].second);
            adj[slots[k].second].push_back(slots[k].first);
        }
    }
    for (int c = 0; c < V; ++c) {
        std::vector<char> seen((size_t)V, 0);
        for (int start = 0; start < V; ++start) {
            if (start == c || seen[start]) continue;
            std::vector<int> stack{start};
            seen[start] = 1;
            bool hasTerminal = false;
            while (!stack.empty()) {
                int x = stack.back();
                stack.pop_back();
                if (x == 0 || x == 1) hasTerminal = true;
                for (int y : adj[x]) {
                    if (y != c && !seen[y]) {
                        seen[y] = 1;
                        stack.push_back(y);
                    }
                }
            }
            if (!hasTerminal) return true;
        }
    }
    return false;
}

bool structureOk(int V, const std::vector<int>& mult, bool allowDead) {
    if (!isConnected(V, mult)) return false;
    if (!allowDead && hasDeadPart(V, mult)) return false;
    return true;
}

std::vector<std::vector<int>> permGroup(int V) {
    std::vector<int> internal;
    for (int v = 2; v < V; ++v) internal.push_back(v);
    std::vector<std::vector<int>> out;
    out.reserve(2 * 1);  // expanded by permutation count below
    for (int swap = 0; swap < 2; ++swap) {
        std::vector<int> pi = internal;  // lexicographic order (itertools)
        do {
            std::vector<int> p((size_t)V);
            p[0] = swap ? 1 : 0;
            p[1] = swap ? 0 : 1;
            for (size_t k = 0; k < pi.size(); ++k) p[pi[k]] = (int)k + 2;
            out.push_back(std::move(p));
        } while (std::next_permutation(pi.begin(), pi.end()));
    }
    return out;
}

std::vector<int> permuteMult(int V, const std::vector<int>& mult,
                             const std::vector<int>& p) {
    const auto& slots = slotList(V);
    std::vector<int> out((size_t)nSlots(V), 0);
    for (size_t k = 0; k < slots.size(); ++k) {
        int a = p[slots[k].first], b = p[slots[k].second];
        if (a > b) std::swap(a, b);
        out[slotIndex(V, a, b)] += mult[k];
    }
    return out;
}

std::vector<int> canonicalMult(int V, const std::vector<int>& mult) {
    std::vector<int> best;
    for (const auto& p : permGroup(V)) {
        std::vector<int> m = permuteMult(V, mult, p);
        if (best.empty() || m < best) best = std::move(m);
    }
    return best;
}

std::vector<std::vector<int>> structureAutomorphisms(int V,
                                                     const std::vector<int>& mult) {
    std::vector<std::vector<int>> out;
    for (const auto& p : permGroup(V)) {
        if (permuteMult(V, mult, p) == mult) out.push_back(p);
    }
    return out;
}

std::string Structure::key() const {
    std::ostringstream ss;
    ss << V << ";";
    for (size_t k = 0; k < mult.size(); ++k) {
        if (k) ss << ",";
        ss << mult[k];
    }
    return ss.str();
}

std::string Structure::serialize() const {
    std::ostringstream ss;
    ss << "V" << V << ":";
    for (size_t k = 0; k < mult.size(); ++k) {
        if (k) ss << ",";
        ss << mult[k];
    }
    return ss.str();
}

std::vector<int> Structure::slotOfInstances() const {
    std::vector<int> out;
    out.reserve((size_t)nEdges());
    for (size_t k = 0; k < mult.size(); ++k)
        out.insert(out.end(), (size_t)mult[k], (int)k);
    return out;
}

Structure makeStructure(int V, std::vector<int> mult, bool canonicalize) {
    if ((int)mult.size() != nSlots(V))
        throw std::invalid_argument("mult has wrong number of slots for V");
    if (canonicalize) mult = canonicalMult(V, mult);
    std::vector<std::vector<int>> aut = structureAutomorphisms(V, mult);
    return Structure{V, std::move(mult), std::move(aut)};
}

std::vector<std::vector<Component>> permuteSlotKeys(
    int V, const std::vector<std::vector<Component>>& keysPerSlot,
    const std::vector<int>& p) {
    const auto& slots = slotList(V);
    std::vector<std::vector<Component>> out(keysPerSlot.size());
    for (size_t k = 0; k < slots.size(); ++k) {
        const auto& keys = keysPerSlot[k];
        if (keys.empty()) continue;
        int a = p[slots[k].first], b = p[slots[k].second];
        if (a > b) std::swap(a, b);
        auto& dst = out[slotIndex(V, a, b)];
        dst.insert(dst.end(), keys.begin(), keys.end());
    }
    for (auto& g : out) std::sort(g.begin(), g.end());
    return out;
}

std::vector<std::vector<Component>> Network::serialize(
    const std::vector<Component>& comps) const {
    const int V = structure.V;
    const auto soi = structure.slotOfInstances();
    std::vector<std::vector<Component>> perSlot((size_t)nSlots(V));
    for (size_t t = 0; t < assign.size(); ++t)
        perSlot[soi[t]].push_back(comps[assign[t]]);
    for (auto& g : perSlot) std::sort(g.begin(), g.end());
    if (structure.aut.size() <= 1) return perSlot;
    std::vector<std::vector<Component>> best = perSlot;
    for (const auto& p : structure.aut) {
        std::vector<std::vector<Component>> m = permuteSlotKeys(V, perSlot, p);
        if (m < best) best = std::move(m);
    }
    return best;
}

bool isSeriesParallel(int V, const std::vector<int>& mult) {
    if (V < 2) return false;
    // ordered count list over occupied slots (insertion order irrelevant to
    // the result; kept deterministic = slot order, new edges appended)
    std::vector<std::pair<std::pair<int, int>, int>> counts;
    auto findSlot = [&counts](int u, int v) -> int {
        for (size_t k = 0; k < counts.size(); ++k)
            if (counts[k].first == std::make_pair(u, v)) return (int)k;
        return -1;
    };
    const auto& slots = slotList(V);
    for (size_t k = 0; k < slots.size(); ++k)
        if (mult[k] > 0) counts.push_back({slots[k], mult[k]});

    std::vector<char> alive((size_t)V, 1);
    while (true) {
        // (a) parallel reduction: collapse every multi-edge to a single edge
        for (auto& kv : counts)
            if (kv.second > 1) kv.second = 1;
        // (b) one series reduction per pass
        std::vector<int> deg((size_t)V, 0);
        std::vector<std::vector<std::pair<int, int>>> inc((size_t)V);
        for (const auto& kv : counts) {
            int a = kv.first.first, b = kv.first.second, m = kv.second;
            deg[a] += m;
            deg[b] += m;
            for (int q = 0; q < m; ++q) {
                inc[a].push_back(kv.first);
                inc[b].push_back(kv.first);
            }
        }
        int target = -1;
        for (int x = 2; x < V; ++x) {
            if (alive[x] && deg[x] == 2) {
                target = x;
                break;
            }
        }
        if (target < 0) break;
        int x = target;
        std::pair<int, int> e1 = inc[x][0], e2 = inc[x][1];
        for (const auto& key : {e1, e2}) {
            int k = findSlot(key.first, key.second);
            counts[k].second -= 1;
        }
        for (const auto& key : {e1, e2}) {
            int k = findSlot(key.first, key.second);
            if (k >= 0 && counts[k].second <= 0) counts.erase(counts.begin() + k);
        }
        alive[x] = 0;
        int others[4] = {e1.first, e1.second, e2.first, e2.second};
        std::vector<int> uniq;
        for (int v : others) {
            if (v != x && std::find(uniq.begin(), uniq.end(), v) == uniq.end())
                uniq.push_back(v);
        }
        if (uniq.size() == 2) {
            int u = std::min(uniq[0], uniq[1]), v = std::max(uniq[0], uniq[1]);
            int k = findSlot(u, v);
            if (k >= 0) counts[k].second += 1;
            else counts.push_back({{u, v}, 1});
        }
        // uniq.size()==1: both edges went to the same partner -> dead pair
    }
    int portCount = 0;
    for (const auto& kv : counts)
        if (kv.first == std::make_pair(0, 1)) portCount = kv.second;
    return portCount == 1 && counts.size() == 1;
}

}  // namespace ng
